File: data/remote/src/main/java/com/chronos/mobile/data/remote/RemoteGithubReleaseRepository.kt
package com.chronos.mobile.data.remote

import android.util.Log
import com.chronos.mobile.domain.GithubReleaseRepository
import com.chronos.mobile.domain.model.GithubRelease
import com.chronos.mobile.domain.result.AppError
import com.chronos.mobile.domain.result.AppResult
import com.chronos.mobile.domain.result.asFailure
import com.chronos.mobile.domain.result.asSuccess
import com.chronos.mobile.domain.result.toAppError
import java.io.IOException
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request

class RemoteGithubReleaseRepository @Inject constructor(
    private val baseClient: OkHttpClient,
) : GithubReleaseRepository {
    private companion object {
        const val TAG = "GithubRelease"
        const val RESPONSE_PREVIEW_LENGTH = 240
    }

    private val json = Json {
        ignoreUnknownKeys = true
    }

    override suspend fun fetchReleaseByTag(
        owner: String,
        repo: String,
        tag: String,
    ): AppResult<GithubRelease> {
        val requestUrl = "https://api.github.com/repos/$owner/$repo/releases/tags/$tag"
        val request = Request.Builder()
            .url(requestUrl)
            .header("Accept", "application/vnd.github+json")
            .header("X-GitHub-Api-Version", "2022-11-28")
            .header("User-Agent", "Chronos-Android-App")
            .build()

        logDebug("Fetching release from $requestUrl")
        return withContext(Dispatchers.IO) {
            try {
                val payload = baseClient.newCall(request).execute().use { response ->
                    val body = response.body.string()
                    if (!response.isSuccessful) {
                        val errorMessage = parseErrorMessage(body)
                            ?: "GitHub release 请求失败：HTTP ${response.code}"
                        logError(
                            "GitHub release request failed: code=${response.code}, message=${response.message}, body=${body.preview()}",
                        )
                        return@withContext AppError.Network(errorMessage).asFailure()
                    }
                    logDebug("GitHub release request succeeded: code=${response.code}, body=${body.preview()}")
                    body
                }
                parseRelease(payload)
            } catch (throwable: Throwable) {
                logError("GitHub release request threw ${throwable::class.java.simpleName}", throwable)
                throwable.toGithubAppError().asFailure()
            }
        }
    }

    private fun parseRelease(raw: String): AppResult<GithubRelease> = runCatching {
        val element = json.parseToJsonElement(raw) as? JsonObject
            ?: throw IllegalStateException("Unexpected release payload")
        val message = element.stringValue("message")
        if (!message.isNullOrBlank()) {
            throw IOException(message)
        }
        GithubRelease(
            tagName = element.stringValue("tag_name").orEmpty(),
            name = element.stringValue("name").orEmpty(),
            publishedAt = element.stringValue("published_at").orEmpty(),
            body = element.stringValue("body").orEmpty(),
            htmlUrl = element.stringValue("html_url").orEmpty(),
        )
    }.fold(
        onSuccess = { it.asSuccess() },
        onFailure = { throwable ->
            when (throwable) {
                is IOException -> AppError.Network(throwable.message ?: "无法获取 GitHub Release 信息，请稍后重试").asFailure()
                else -> AppError.DataFormat("GitHub release 数据格式错误").asFailure()
            }
        },
    )

    private fun JsonObject.stringValue(key: String): String? =
        this[key]?.jsonPrimitive?.contentOrNull

    private fun parseErrorMessage(raw: String): String? = runCatching {
        val element = json.parseToJsonElement(raw)
        (element as? JsonObject)?.stringValue("message")
    }.getOrNull()

    private fun String.preview(): String =
        replace('\n', ' ').take(RESPONSE_PREVIEW_LENGTH)

    private fun logDebug(message: String) {
        runCatching { Log.d(TAG, message) }
    }

    private fun logError(message: String, throwable: Throwable? = null) {
        runCatching {
            if (throwable == null) {
                Log.e(TAG, message)
            } else {
                Log.e(TAG, message, throwable)
            }
        }
    }

    private fun Throwable.toGithubAppError(): AppError {
        val mapped = toAppError()
        return if (mapped is AppError.Unknown) {
            AppError.Network("无法获取 GitHub Release 信息，请检查网络后重试")
        } else {
            mapped
        }
    }
}

File: src/lib/client/github/remote-github-release-repository.ts
import type { GithubRelease } from '$lib/models/auth';
import type { GithubReleaseRepository } from '$lib/domain/interfaces/github-release-repository';
import { failure, success, type AppResult } from '$lib/domain/result/app-result';
import { AppError } from '$lib/domain/result/app-error';

const USER_AGENT = 'Chronos-PWA';

export function createRemoteGithubReleaseRepository(): GithubReleaseRepository {
	return {
		async fetchReleaseByTag(owner, repo, tag) {
			const requestUrl = `https://api.github.com/repos/${owner}/${repo}/releases/tags/${tag}`;
			try {
				const response = await fetch(requestUrl, {
					headers: {
						Accept: 'application/vnd.github+json',
						'X-GitHub-Api-Version': '2022-11-28',
						'User-Agent': USER_AGENT
					}
				});
				const body = await response.text();
				if (!response.ok) {
					const message =
						parseErrorMessage(body) ?? `GitHub release 请求失败：HTTP ${response.status}`;
					return failure(AppError.network(message));
				}
				return parseRelease(body);
			} catch (error) {
				return failure(
					AppError.network(
						error instanceof Error
							? error.message
							: '无法获取 GitHub Release 信息，请检查网络后重试'
					)
				);
			}
		}
	};
}

function parseRelease(raw: string): AppResult<GithubRelease> {
	try {
		const element = JSON.parse(raw) as Record<string, unknown>;
		const message = stringValue(element, 'message');
		if (message) {
			return failure(AppError.network(message));
		}
		return success({
			tagName: stringValue(element, 'tag_name') ?? '',
			name: stringValue(element, 'name') ?? '',
			publishedAt: stringValue(element, 'published_at') ?? '',
			body: stringValue(element, 'body') ?? '',
			htmlUrl: stringValue(element, 'html_url') ?? ''
		});
	} catch {
		return failure(AppError.dataFormat('GitHub release 数据格式错误'));
	}
}

function stringValue(element: Record<string, unknown>, key: string): string | undefined {
	const value = element[key];
	return typeof value === 'string' ? value : undefined;
}

function parseErrorMessage(raw: string): string | undefined {
	try {
		const element = JSON.parse(raw) as Record<string, unknown>;
		return stringValue(element, 'message');
	} catch {
		return undefined;
	}
}
